### PR TITLE
Should be easier to use, with dependencies from Homebrew and mimic build conf. Documentation enabled by default. 

### PR DESCRIPTION
**Pull request by seivan**
This will allow `asdf-erlang` & `asdf-elixir` to be used out of the box and be much easier for majority of macOS users.
Avoiding using esotoric configuration and environment variables to keep track for various plugins should be a focus.

So,  at least for macOS users asdf-erlang should try to use as much as possible from Homebrew for building, with sane configuration defaults (mimic Homebrew) and make documentation enabled by default

* Will use existing kerl from `$PATH` (like installed by Homebrew) or fetch specific version when setting `ASDF_KERL_VERSION`.

* Will use git to clone `kerl` quitly and checkout versioned tags instead of downloading from Github CDN.

* Enable shell docs out of the box by default and allow opt out using `KERL_BUILD_DOCS="no"` - technically any value will opt out, kerl just checks if it's set, and not its value. Since we're opting in by default, we will check `!= "yes"` for opt out.

* Will enable prebuilt binaries for macOS supporting (supported versions of Erlang will of course vary) 
  * Yosemite
  * El Capitan
  * Sierra
  * High Sierra
  * Mojave
  * Catalina
  * Big Sur
  * ARM Big Sur

This allows installing Erlang in matter of seconds versus 10 minutes on slower machines. 
Can optout of prebuilt binaries for macOS with `ASDF_ERLANG_FROM_SOURCE=YES`
These will also opt out: 

* `ASDF_INSTALL_TYPE="ref"`
* `KERL_CONFIGURE_OPTIONS`
* `OTP_GITHUB_URL`

Failing to find proper binaries will default to installing from source. 

* No longer symlinks binaries from globs, instead added a `list-bin-paths` that outputs all the directories with binaries.  Preferably this should have been handled by asdf itself, but will have to do for now. 

* Will come with out of the box settings for `KERL_CONFIGURE_OPTIONS`: 
	* `--disable-debug`
	* `--disable-silent-rules`
	* `--enable-dynamic-ssl-lib`
	* `--enable-hipe`
	* `--enable-sctp`
	* `--enable-shared-zlib`
	* `--enable-smp-support`
	* `--enable-threads`
	* `--enable-wx`
	* `--without-javac`
	* macOS only:
		* `--with-ssl=$(brew --prefix openssl)`
		* `--enable-darwin-64bit`
		* `--enable-kernel-poll`
		* `--with-dynamic-trace=dtrace`

* macOS only: Unless custom `KERL_CONFIGURE_OPTIONS`, asdf-erlang will default to using Homebrew for following dependencies:
 	- `openssl`
 	- `wxmac`
 	- Building from source:
	  - `autoconf`
 	  - `libtool`

Keep in mind, this treats `KERL_BASE_DIR` and `kerlc` as ephemeral by keeping it in all in `tmp`.
You can override this by setting `ASDF_KERL_BASE_DIR` and it will set `KERL_CONFIG=$ASDF_KERL_BASE_DIR/.kerlc`

That being said, as a user of asdf-erlang, I'd assume you're not really opting to interact with it by default. So it would make sense to treat it as ephemeral resource more than a configuration to keep around. 

Using default `KERL_CONFIGURE_OPTIONS` if unset. 
For `macOS` this contains additional checks, where it depends on Homebrew and packages installed by it. 

So if a macOS user wants to opt out of dependencies used by Homebrew, they would need to set their own `KERL_CONFIGURE_OPTIONS`. 

This is similar to changes in https://github.com/asdf-vm/asdf-ruby/pull/205 I am trying to keep the logic and folder organization the same, since I'm maintaining both of these forks for myself, hence moving `utils.sh` out to `lib/utils.sh`, also it feels cleaner not to have non-asdf commands under `bin/`

### CODE REVIEW

**Comment by seivan** [comment]
I think your CI is broken.
The macOS one fails because `wxmac` requirement which should be standard for macOS and the Linux one fails for some other reason even though the installation succeeded. 
https://travis-ci.org/github/asdf-vm/asdf-erlang/jobs/762870648#L132

File: lib/utils.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  if [ "$(uname -s)" = "Darwin" ]; then
+    osx_num=$(sw_vers -productVersion| awk -F '[.]' '{print $2}')
+    OSX_MARKETING=(
+        ["10"]="yosemite"
+        ["11"]="el_capitan"
+        ["12"]="sierra"
+        ["13"]="high_sierra"
+        ["14"]="mojave"
+        ["15"]="catalina"
+        ["16"]="big_sur"
```
Comment (wojtekmach) [inline lib/utils.sh:66]: this only works when `SYSTEM_VERSION_COMPAT=1` is set:

```
$ SYSTEM_VERSION_COMPAT=1
$ sw_vers -productVersion
10.16

$ unset SYSTEM_VERSION_COMPAT
$ sw_vers -productVersion
11.2.3
```

I think we could set this env var for the lifetime of this function and then possibly clean it up but leaving it for the lifetime of the whole install should be fine too.

File: bin/install [hunk 2]
```diff
@@ -1,54 +1,144 @@
...
-  echo "Erlang $ASDF_INSTALL_VERSION has been installed. Activate globally with:"
-  echo
-  echo "    asdf global erlang $asdf_activation_version"
-  echo
-  echo "Activate locally in the current folder with:"
-  echo
-  echo "    asdf local erlang $asdf_activation_version"
-  echo
+  # Replace the homebrew URL with something compatible with asdf
+  grep -rl --null "$replacement_pattern" "$ASDF_DOWNLOAD_PATH" | xargs -0 sed -i '' "s:$replacement_pattern:$ASDF_INSTALL_PATH:g"
```
Comment (wojtekmach) [inline bin/install:63]: This failed for me with:

```
sed: /Users/wojtek/.asdf/downloads/erlang/23.3/bin/erl: in-place editing only works for regular files
```

and it's because it's a symlink:

```
~% ls -l /Users/wojtek/.asdf/downloads/erlang/23.3/bin/erl
lrwxr-xr-x  1 wojtek  staff  21 Mar 22 13:27 /Users/wojtek/.asdf/downloads/erlang/23.3/bin/erl -> ../lib/erlang/bin/erl
```

**Review by wojtekmach — COMMENTED** [review]
This is great, I'm really looking forward to having this in asdf-erlang.

File: lib/utils.sh (hunk 1 above)
Comment (seivan) [reply lib/utils.sh:66]: Though I am not sure how subshells work, I don't think there is a need to clean it up if we're enabling it for the lifetime of the function. 

Alternatives: 
* Work with both `2` and `16`
* Toggle before calling `sw_vers -productVersion` and Toggle back. 

I think just setting it for the subshell might be good enough, what's your call?

File: bin/install (hunk 2 above)
Comment (seivan) [reply bin/install:63]: > This failed for me with:
> 
> ```
> sed: /Users/wojtek/.asdf/downloads/erlang/23.3/bin/erl: in-place editing only works for regular files
> ```
> 
> and it's because it's a symlink:
> 
> ```
> ~% ls -l /Users/wojtek/.asdf/downloads/erlang/23.3/bin/erl
> lrwxr-xr-x  1 wojtek  staff  21 Mar 22 13:27 /Users/wojtek/.asdf/downloads/erlang/23.3/bin/erl -> ../lib/erlang/bin/erl
> ```

Those symlinks are thrown away at the end anyway, so I'll just be more specific with the grep.

File: lib/utils.sh (hunk 1 above)
Comment (wojtekmach) [reply lib/utils.sh:66]: yeah I think setting it just for the subshell sounds good

File: bin/install (hunk 2 above)
Comment (seivan) [reply bin/install:63]: That being said, did not fail for me, running Catalina without coreutils. 
Sigh, hate these things. Anyway pushed some changes, feel free to try again.

Comment (wojtekmach) [reply bin/install]: Yup, it works!

```
~% asdf install erlang 23.3
Installing erlang 23.3 for big_sur
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
100  101M  100  101M    0     0  29.2M      0  0:00:03  0:00:03 --:--:-- 53.7M
~% asdf global erlang 23.3
~% erl
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Eshell V11.2  (abort with ^G)
1>
```

Comment (seivan) [reply bin/install]: Just to make sure, could you pop into elixir and run `h :ets.lookup`?

Comment (seivan) [reply bin/install]: Resolving this since it worked. 
Prebuilt binaries come with repl docs on newer OTP so it will probably work. Building from source will always opt into docs but was tricky to test configs as it took ~15 min between installs, lol.

Comment (wojtekmach) [reply bin/install]: Yup, all is well:

```
~% asdf current
elixir          1.11.3-otp-23   /Users/wojtek/.tool-versions
erlang          23.3            /Users/wojtek/.tool-versions
~% iex
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Interactive Elixir (1.11.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> h :ets.lookup

                                    lookup/2

  @spec lookup(tab, key) :: [object]
        when tab: tab(), key: term(), object: tuple()

Returns a list of all objects with key Key in table Tab.

  • For tables of type set, bag, or duplicate_bag, an object is returned only if
    the specified key matches the key of the object in the table.

  • For tables of type ordered_set, an object is returned if the specified key
    compares equal to the key of an object in the table.
(...)
```

Comment (seivan) [reply bin/install]: Glad to hear! Thanks for the help dude.

Comment (wojtekmach) [reply bin/install]: I did some more testing and it unfortunately doesn't work, crypto is broken due to paths.

```
~% iex
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Interactive Elixir (1.11.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> :crypto.versions()

01:30:26.637 [error] Unable to load crypto library. Failed with error:
":load_failed, Failed to load NIF library: 'dlopen(/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so, 2): Library not loaded: @@HOMEBREW_PREFIX@@/opt/openssl@1.1/lib/libcrypto.1.1.dylib
  Referenced from: /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
  Reason: image not found'"
OpenSSL might not be installed on this system.


01:30:26.644 [warn]  The on_load function for module crypto returned:
{:error,
 {:load_failed,
  'Failed to load NIF library: \'dlopen(/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so, 2): Library not loaded: @@HOMEBREW_PREFIX@@/opt/openssl@1.1/lib/libcrypto.1.1.dylib\n  Referenced from: /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so\n  Reason: image not found\''}}

```



```
~% otool -L /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so:
	@@HOMEBREW_PREFIX@@/opt/openssl@1.1/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
```

Comment (seivan) [reply bin/install]: Sigh, I guess there is no choice than to wait until OPT gets support for different ROOT_DIR. 
Will still investigate a way to fix this, but if I can't find one, I'll get rid of the prebuilt binary support.

Comment (wojtekmach) [reply bin/install]: I remember playing with `install_name_tool` [1] for stuff like that, curious if brew has post-install scripts using something similar.

[1] https://blog.krzyzanowskim.com/2018/12/05/rpath-what/

Comment (seivan) [reply bin/install]: It certainly does  https://github.com/Homebrew/brew/blob/e35e5f13e81ca114c18b09bec9320e5b00fe021e/Library/Homebrew/formula.rb#L1048

Comment (seivan) [reply bin/install]: @wojtekmach Ok, so I added a tempfix just to make it work for me. 
Also are you sure it's `:crypto.versions()` and not `:crypto.version()`? 
I pushed this temp fix, but it relies on hard coded values, might not work for you.

If you can help to find anything that's dynamically linked, I can probably add code to fix most of them once I figure out how to not hardcode versions in the paths. 

Will be a hard dependency on homebrew  even with prebuilt binaries, but that will have to do.

Comment (seivan) [reply bin/install]: Ok, so latest push no longer hard codes the values. But depends on homebrew and install_name_tool
```
$ where install_name_tool 
/usr/bin/install_name_tool
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool
```

Comment (wojtekmach) [reply bin/install]: oops, yup, should be `:crypto.version()` or `:ssl.versions()`, I always mix them up.

I gave temp fix a shot and unfortunately got this:

```
~% asdf install erlang 23.3
Installing erlang 23.3 for big_sur
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  101M  100  101M    0     0  36.4M      0  0:00:02  0:00:02 --:--:-- 63.3M


** emphasis mine **
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/crypto-4.9/priv/lib/crypto.so
**


~% asdf global erlang 23.3
~% iex
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Interactive Elixir (1.11.3) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> :crypto.version
zsh: killed     iex
```

Comment (seivan) [reply bin/install]: Haha, sounds like a Big Sur change. 
Considering disabling Big Sur (by removing it from mac_os_marketing_name) until there is a better way of doing this. 

Which commit was this? 
I pushed 89a2ccd but that doesn't anything but hard coded values. 

But this kinda does make sense if we're changing a linked value then obviously it'll get disabled. 
Does work on Catalina, but I understand that's something that has to be tackled as more people move to Big Sur.

Comment (seivan) [reply bin/install]: >  If you use a custom workflow involving tools that modify a binary after linking (e.g. strip or install_name_tool) you might need to manually call codesign(1) as an additional build phase to properly ad-hoc sign your binary. These new signatures are not bound to the specific machine that was used to build the executable, they can be verified on any other system and will be sufficient to comply with the new default code signing requirement on Macs with Apple silicon.

I am wondering if we can call code sign after changing the loaded lib?

Comment (seivan) [reply bin/install]: @wojtekmach Could you run `codesign -v -vvvv -ignore-resources crypto.so ` and see what you get?

Comment (wojtekmach) [reply bin/install]: no dice:

```
% codesign -v -vvvv -ignore-resources /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so: invalid signature (code or signature have been modified)
In architecture: arm64
```

Comment (seivan) [reply bin/install]: No that's good. We now it's invalid, because bin/install changed it. 
Now, go ahead and sign it :)

Comment (wojtekmach) [reply bin/install]: sorry I'm not familiar with this, how do I sign stuff?

Comment (seivan) [reply bin/install]: Hold on a second, technically this sorta stuff shouldn't work on my mac either, but I am not sure if Big Sur has more safe rails than Catalina. 

Do you have this enabled?
image

Comment (wojtekmach) [reply bin/install]: I tried the following:

```
~% codesign -v -vvvv -ignore-resources /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so: invalid signature (code or signature have been modified)
In architecture: arm64
~% codesign -s - /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so: is already signed
```

Comment (seivan) [reply bin/install]: try `codesign -s - -f /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so`

Comment (seivan) [reply bin/install]: Ah yeah, it seems like Big Sur has more stuff to take into account

> the system ships with a built-in dynamic linker cache of all system-provided libraries. As part of this change, copies of dynamic libraries are no longer present on the filesystem. Code that attempts to check for dynamic library presence by looking for a file at a path or enumerating a directory will fail. Instead, check for library presence by attempting to dlopen() the path, which will correctly check for the library in the cache. (62986286)

Comment (wojtekmach) [reply bin/install]: I got:

```
~% codesign -s - -f /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so: replacing existing signature
/Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so: the codesign_allocate helper tool cannot be found or used
```

but I think I screwed up my Xcode/command line tools installation while checking out other stuff.

I'm done for now, will check in tomorrow :) Thank you for the help and creating the PR in the first place!

Comment (seivan) [reply bin/install]: Try this  (gonna crash soon) 

```
cp -r crypto.so backup_crypto.so
codesign -s - -f -vvvvvv backup_crypto.so
rm -rf crypto.so
mv backup_crypto.so crypto.so
```
Related https://github.com/Homebrew/brew/issues/9082

Comment (wojtekmach) [reply bin/install]: yup, that did it!

```
iex(4)> :inets.start; :ssl.start; :httpc.request("https://github.com")
{:ok,
 {{'HTTP/1.1', 200, 'OK'}, ...
```

Comment (seivan) [reply bin/install]: > I got:
```
 /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so: the codesign_allocate helper tool cannot be found or used
 ```
> 
> but I think I screwed up my Xcode/command line tools installation while checking out other stuff.

This is a bug, you need to copy it first. Good night!

Comment (seivan) [reply bin/install]: I am making the assumption that macOS users have `codesign` & `install_name_tools` installed via Xcode CLI. Those would be required from xcode CLI anyway.

So did `codesign -s - -f -vvvvvv <file>` work for you without copying or with copying?

Comment (wojtekmach) [reply bin/install]: It only worked when copying.

Here's full log:

```
~% asdf install erlang 23.3
Installing erlang 23.3 for big_sur
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
100  101M  100  101M    0     0  18.5M      0  0:00:05  0:00:05 --:--:-- 24.7M
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/crypto-4.9/priv/lib/crypto.so

~% ~/.asdf/installs/erlang/23.3/bin/erl -noshell -eval 'io:format("~p~n", [crypto:version()]), halt()'
zsh: killed     ~/.asdf/installs/erlang/23.3/bin/erl -noshell -eval
~%
~% cp ~/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so backup_crypto.so
~% codesign -s - -f -vvvvvv backup_crypto.so
backup_crypto.so: replacing existing signature
backup_crypto.so: signed Mach-O thin (arm64) [backup_crypto-55554944ed15fd397c9c365aa097ec765098dbd6]
~% mv backup_crypto.so ~/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so
override r--r--r--  wojtek/staff for /Users/wojtek/.asdf/installs/erlang/23.3/lib/crypto-4.9/priv/lib/crypto.so? (y/n [n]) y

~% ~/.asdf/installs/erlang/23.3/bin/erl -noshell -eval 'io:format("~p~n", [crypto:version()]), halt()'
"4.9"
```

Thanks again!

**Comment by seivan** [comment]
Ok! That looks ridiculous enough to work. 

Could you check that :observer.start works?
I know it’s not dynamically linked but just wanna be sure we’re covering all bases. 

I’ll push a code sign fix for just Big Sur:

Update linker flag
Copy,
Code sign
Delete original 
Copy back

That’ll do?

If this works I can build one for Ruby as well. 


Sent from an iPhone on the go.

File: bin/install (hunk 2 above)
Comment (seivan) [reply bin/install]: Ok! That looks ridiculous enough to work. 

Could you check that :observer.start works?
I know it’s not dynamically linked but just wanna be sure we’re covering all bases. 

I’ll push a code sign fix for just Big Sur:

Update linker flag
Copy,
Code sign
Delete original 
Copy back

That’ll do?

If this works I can build one for Ruby as well. 

Edit: just noticed you’re getting the arm version, correct?
I am wondering if this is necessary for just Big Sur or Big Sur arm. We’ll start with both archs for Big Sur 

Sent from an iPhone on the go.

**Comment by wojtekmach** [comment]
We're gonna have to do similar workaround for wx, I got this error:

```
~/.asdf/installs/erlang/23.3/bin/erl
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Eshell V11.2  (abort with ^G)
1> observer:start().
=ERROR REPORT==== 30-Mar-2021::15:54:59.777145 ===
WX Failed loading "wxe_driver"@"/Users/wojtek/.asdf/installs/erlang/23.3/lib/wx-1.9.3/priv"

{error,{{load_driver,"dlopen(/Users/wojtek/.asdf/installs/erlang/23.3/lib/wx-1.9.3/priv/wxe_driver.so, 2): Library not loaded: @@HOMEBREW_PREFIX@@/opt/wxmac/lib/libwx_osx_cocoau_stc-3.0.dylib\n  Referenced from: /Users/wojtek/.asdf/installs/erlang/23.3/lib/wx-1.9.3/priv/wxe_driver.so\n  Reason: image not found"},
        [{wxe_server,start,1,[{file,"wxe_server.erl"},{line,65}]},
         {wx,new,1,[{file,"wx.erl"},{line,115}]},
         {observer_wx,init,1,[{file,"observer_wx.erl"},{line,107}]},
         {wx_object,init_it,6,[{file,"wx_object.erl"},{line,404}]},
         {proc_lib,init_p_do_apply,3,
                   [{file,"proc_lib.erl"},{line,226}]}]}}
```

**Comment by wojtekmach** [comment]
Yes, I have wxmac installed through brew. And yeah I can test it on a x64 machine, I'll try it later today.

**Comment by seivan** [comment]
@wojtekmach I think I found why it worked for me. 
```
otool -L priv/wxe_driver.so 
priv/wxe_driver.so:
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 162.0.0)
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 23.0.0)
	/System/Library/Frameworks/AudioToolbox.framework/Versions/A/AudioToolbox (compatibility version 1.0.0, current version 1000.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
``
You don't have those on ARM_big_sur, but x64 Big Sur does. 

So if we're updating linker flags for wxmac, it should just be for ARM.

**Comment by seivan** [comment]
@wojtekmach Latest push will do this

Introducing `macos_update_linked_path()` that 
* Checks for `@@HOMEBREW_PREFIX@@` on glob paths - so far hardcoded to `wxmac` & `crypto`  like `"$ASDF_DOWNLOAD_PATH/lib/erlang/lib/"crypto-*"/priv/lib/crypto.so"` , can make it dynamic in the future to look at *ALL* libs.

* Runs `otool -L` to figure out what paths need to change.
* Uses `install_name_tool -change` with `$(brew --prefix)`
* Copies the file before running `codesign` [because of bug](https://github.com/Homebrew/brew/blob/e35e5f13e81ca114c18b09bec9320e5b00fe021e/Library/Homebrew/formula.rb#L1048)
* Moves file back. 

```
  # Link Crypto to brew prefix
  macos_update_linked_path "$ASDF_DOWNLOAD_PATH/lib/erlang/lib/"crypto-*"/priv/lib/crypto.so"
  # Link Wxmac to brew prefix
  macos_update_linked_path "$ASDF_DOWNLOAD_PATH/lib/erlang/lib/"wx-*"/priv/wxe_driver.so"
```

I ran this on Catalina and it had no issues, so we'll just leave it in there for all macOS major versions.

**Comment by seivan** [comment]
@wojtekmach 
Ok so I removed a lot of logging. 
Also , changed so we're only running `install_name_tools` once instead of looping over for each change. 
This mean at most you should get two outputs
```
.asdf/downloads/erlang/23.3/lib/erlang/lib/crypto-4.9/priv/lib/crypto.so.backup: signed Mach-O thin (x86_64) [crypto.so]
.asdf/downloads/erlang/23.3/lib/erlang/lib/wx-1.9.3/priv/wxe_driver.so.backup: signed Mach-O thin (x86_64) [wxe_driver.so]
```

Ok, I think I'm done for now.
On my end, the only issue I had was Crypto, wxmac never needed any changes on my side, but even with new path updates and code sign it works.

**Comment by seivan** [comment]
@wojtekmach Are we done here?

**Comment by wojtekmach** [comment]
```
~% asdf install erlang 23.3
Looking for prebuilt erlang 23.3 on arm64_big_sur.
Checking Homebrew for dependencies.
Downloading https://bintray.com/homebrew/bottles/download_file?file_path=erlang-23.3.arm64_big_sur.bottle.tar.gz.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  101M  100  101M    0     0  31.8M      0  0:00:03  0:00:03 --:--:-- 50.0M
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/crypto-4.9/priv/lib/crypto.so
/Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/crypto-4.9/priv/lib/crypto.so.backup: replacing existing signature
/Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/crypto-4.9/priv/lib/crypto.so.backup: signed Mach-O thin (arm64) [crypto.so]
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/wx-1.9.3/priv/wxe_driver.so
/Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/wx-1.9.3/priv/wxe_driver.so.backup: replacing existing signature
/Users/wojtek/.asdf/downloads/erlang/23.3/lib/erlang/lib/wx-1.9.3/priv/wxe_driver.so.backup: signed Mach-O thin (arm64) [wxe_driver.so]
~% ~/.asdf/installs/erlang/23.3/bin/erl
cryErlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [dtrace]

Eshell V11.2  (abort with ^G)
1> crypto:version().
"4.9"
2> observer:start().
ok
```

:+1:

**Comment by wojtekmach** [comment]
@seivan what happened? :)

**Comment by seivan** [comment]
@wojtekmach 
*TLDR;* I am no longer using main asdf. I appreciate your help and feedback! 
I restored the branch if you need it. 

I am doing prebuilt (macOS) support for more languages, and a lot of this logic would do better outside of the plugins. 
Went from copying code, to using submodules and eventually settled on setting the logic outside of plugins for prebuilts to use on other languages like Ruby, Node, & etc.

**Comment by wojtekmach** [comment]
Out of curiosity, by keeping logic outside of the plugins you mean inside asdf vm core or outside asdf? In any case, I and I'm sure many people really appreciate your effort and showing the way to implement this, thank you!

**Comment by seivan** [comment]
> Out of curiosity, by keeping logic outside of the plugins you mean inside asdf core or outside asdf?

That's the part that bothers me, so *right now* it's a submodule used by ruby, node and erlang. 
It has a lot shared components to curl tarballs, extract, replace linked libs, replace code sign and etc. 

Core does not afaik(!) know deal with submodules, and I'm on the fence wether or not it's a bad idea, but that's the approach I am going with now. 


There's also silly util functions like getting plugin dir, check for necessary dependencies and (for macOS) offer an approach to fetch them -using using Homebrew. Similar to how `ruby-install` does it. 


>  In any case, I and I'm sure many people really appreciate your effort and showing the way to implement this, thank you!

We'll see when it's done and I can I can show it. 

So far all three languages work fine, even Ruby's quirk of hard coding gem loading paths in the binary (based on Brew) - we override this by using [exec-env](https://github.com/asdf-vm/asdf-ruby/pull/205/files#diff-bd7e16b2f036811f6a156b2f59efc2e7621107b5fc9240c020fdd8caf127095eR6-R8) - right now hard-coded for macOS. It will remain so, but will check it's macOS before doing so later on.

**Comment by wojtekmach** [comment]
Gotcha, makes sense to share the behaviour somehow so that plugins can easily use it.

**Comment by Stratus3D** [comment]
@seivan @wojtekmach I am applying some of these changes in https://github.com/asdf-vm/asdf-erlang/pull/276. The main Homebrew changes here won't be accepted - at least for now - but I figured I'd let you know some of this is going to end up in `master`.
